django/utils/choices.py

```diff
@@ -0,0 +1,51 @@
+from collections.abc import Callable, Iterable, Iterator, Mapping
+
+from django.db.models.enums import ChoicesMeta
+from django.utils.functional import Promise
+
+
+class ChoiceIterator:
+    """Base class for lazy iterators for choices."""
+
+
+class CallableChoiceIterator(ChoiceIterator):
+    """Iterator to lazily normalize choices generated by a callable."""
+
+    def __init__(self, choices_func):
+        self.choices_func = choices_func
+
+    def __iter__(self):
+        yield from normalize_field_choices(self.choices_func())
+
+
+def normalize_field_choices(value, *, depth=0):
+    """Normalizes choices values consistently for fields and widgets."""
+
+    match value:
+        case ChoicesMeta():
+            # Choices enumeration helpers already produce output in canonical form.
+            return value.choices
+        case ChoiceIterator() | Promise() | bytes() | str():
+            # Avoid expanding lazy choice iterators too soon by returning early.
+            # String-like types are iterable, so return early to avoid iterating.
+            return value
+        case Callable() if depth == 0:
+            # If at the top level, wrap callables to be evaluated lazily.
+            return CallableChoiceIterator(value)
+        case Mapping():
+            value = value.items()
+        case Iterator():
+            pass
+        case Iterable() if not any(isinstance(x, (Promise, bytes, str)) for x in value):
+            pass
+        case _:
+            return value
+
+    try:
+        # Recursive call to convert possibly nested values to a list of 2-tuples.
+        return [(k, normalize_field_choices(v, depth=depth + 1)) for k, v in value]
+    except (ValueError, TypeError):
+        # Return original value for the system check to raise if it has items that are:
+        # - Not iterable, i.e. cannot unpack non-iterable <type> object (TypeError)
+        # - Not 2-tuples, i.e. not enough / too many values to unpack (ValueError)
+        return value
```

Comment:
This is quite self-explanatory with the comments. We need to return the original incorrect values so that the model system checks are raised correctly...

Having said that, it does make me wonder whether we ought to raise something out here and simplify those checks to catch appropriate exceptions, rather than iterating all over again in the checks? 🤔

Reply:
> …to raise something…

Thinking about this more, that wouldn't work because we don't want this to cause a crash. Normalisation will occur well before the system checks run… Still, it'd be more consistent and maintainable to share the implementation somehow.

Reply:
Resolving this comment. We can always consider simplifying the normalisation vs system check stuff as a follow up if we so desire.